File: src/dataframe/rust/stats/distributions/beta.test.ts
#!/usr/bin/env -S deno run --allow-read --allow-run

import { callR, printResults, runComparison } from "./test-helpers.ts";

async function main() {
  console.log("Testing R vs Rust beta functions...\n");

  const testCases = [
    { func: "dbeta", distribution: "beta", args: [0.5, 2, 3, 0] },
    { func: "pbeta", distribution: "beta", args: [0.5, 2, 3, 1, 0] },
    { func: "qbeta", distribution: "beta", args: [0.6875, 2, 3, 1, 0] },
  ];

  const results = await runComparison(testCases);
  printResults(results);

  // Test rbeta separately since it's not available in Rust interface
  console.log("\nR-only tests:");
  try {
    const rbetaResult = await callR("rbeta", "beta", 1, 2, 3);
    console.log(`rbeta(2, 3) = ${rbetaResult}`);
  } catch (error) {
    console.log(`rbeta: Error - ${(error as Error).message}`);
  }
}

if (import.meta.main) {
  main().catch(console.error);
}

File: packages/dataframe/rust/stats/distributions/beta.test.R
#!/usr/bin/env Rscript

# Beta distribution test script
# Usage: Rscript beta.test.R <function> <x_or_p> <shape1> <shape2> [lower_tail] [log_p]

source("test-helpers.R")

# Define beta distribution functions
beta_density <- function(x, shape1, shape2, log = FALSE) {
  dbeta(x, shape1, shape2, log = log)
}

beta_cumulative <- function(x, shape1, shape2, lower.tail = TRUE, log.p = FALSE) {
  pbeta(x, shape1, shape2, lower.tail = lower.tail, log.p = log.p)
}

beta_quantile <- function(p, shape1, shape2, lower.tail = TRUE, log.p = FALSE) {
  qbeta(p, shape1, shape2, lower.tail = lower.tail, log.p = log.p)
}

beta_random <- function(n, shape1, shape2) {
  rbeta(n, shape1, shape2)
}

# Run the test
run_distribution_test(beta_density, beta_cumulative, beta_quantile, beta_random)
